driver: bluetooth: HCI: add Ambiq Apollo3 BLE support

aaronyegx opened the pull request:
This PR implements the SPI based HCI driver for Ambiq Apollo3 Blue/Apollo3 Plus Blue.  Apollo3 Blue/Apollo3 Plus Blue BLE integrates one core for running BLE controller only. The MCU can work as BLE host and talk to the integrated BLE controller via SPI based BLEIF module. The SPI BLEIF pins are not exposed from MCU and the new implemented SPI BLEIF driver is added for HCI.

Test passed on the following samples:
-samples/bluetooth/peripheral
-samples/bluetooth/peripheral_hr
-samples/bluetooth/peripheral_ht
-samples/bluetooth/peripheral_esp
-samples/bluetooth/peripheral_dis
-samples/bluetooth/peripheral_gatt_write
-samples/bluetooth/peripheral_sc_only
-samples/bluetooth/central
-samples/bluetooth/central_hr
-samples/bluetooth/central_ht
-samples/bluetooth/central_gatt_write
-samples/bluetooth/broadcaster
-samples/bluetooth/broadcaster_multiple
-samples/bluetooth/beacon
-samples/bluetooth/observer
-samples/bluetooth/scan_adv
-samples/bluetooth/mtu_update
-samples/bluetooth/eddystone
-samples/bluetooth/ibeacon
-samples/bluetooth/hci_uart
-tests/bluetooth/init

RichardSWheatley commented on boards/ambiq/apollo3p_evb/Kconfig.defconfig:
```diff
@@ -0,0 +1,41 @@
+
+if BOARD_APOLLO3P_EVB
+
+if BT
```
Should this section go in the SOC since the chip is BLE?

aaronyegx replied:
I moved the default configurations of BT to SOC kconfig. Thanks.

RichardSWheatley reviewed (dismissed):
As far as the Ambiq stuff goes, this looks good.
I will let the other BLE people discuss that section.

aaronyegx commented:
@jhedberg it seems you have another PR for HCI driver API change https://github.com/zephyrproject-rtos/zephyr/pull/72323 and HCI Ambiq driver will also be modified (not update yet). So should this PR for Apollo3 support go first or after that PR getting done?

jhedberg commented:
> @jhedberg it seems you have another PR for HCI driver API change #72323 and HCI Ambiq driver will also be modified (not update yet). So should this PR for Apollo3 support go first or after that PR getting done?

It's likely that the HCI API rewrite will only get merged after the next release, so in that sense any new drivers (using the old API) should be able to be merged before that. For now, I also have `#if DT_HAS_CHOSEN(zephyr_bt_hci)` blocks in my PR in order to support drivers using the old API, but those might go away before my PR is merged.

aaronyegx committed:
drivers: spi: create Ambiq SPI BLEIF driver

Some Ambiq Apollox Blue SOC (e.g. Apollo3 Blue) uses internal designed
BLEIF module which is different from the general IOM module for SPI
transceiver. The called HAL API will also be independent. This driver is
implemented for the BLEIF module usage scenarios.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
boards: arm: apollo3p_evb: create BLEIF instance

This commits add the BLEIF instance which is compatible with
"ambiq,spi-bleif" on Ambiq apollo3p_evb and apollo3_evb.
Also creates the default pinctrl for the defined instance.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
drivers: bluetooth: hci: add Ambiq Apollo3 Blue SOC support

This commit add the SPI-based HCI support for the Ambiq Apollo3 Blue
SOC (e.g. Apollo3 Blue Plus, Apollo3 Blue) support.
Also correct the dependency of necessary peripheral.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
drivers: bluetooth: hci: correct the Ambiq Apollox support BT feature

The BLE controller of some Ambiq Apollox Blue SOC may have issue to
report the expected supported features bitmask successfully, thought the
features are actually supportive. Need to correct them before going to
the host stack.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
drivers: bluetooth: hci: add retry mechanism for Apollo SPI busy scenario

The controller may be unavailable to receive packets because it is busy
on processing something or have packets to send to host. Need to free the
SPI bus and wait some moment to try again.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
dts: arm: ambiq: add bt-hci subnode for Apollo3 Blue SOC

This commit defines the bt-hci subnode under the bleif node on
Ambiq Apollo3 Blue and Apollo3 Blue Plus SOC.
Also add the default configurations for Bluetooth feature on Ambiq
apollo3_evb and apollo3p_evb.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx commented:
@jhedberg Would you mind taking a look at this PR when you are available? Thank you.

jhedberg commented:
@aaronyegx approved from my side, but I'll likely need some guidance on how to best convert this to the new HCI API (especially the DTS parts of it).

aaronyegx commented:
> @aaronyegx approved from my side, but I'll likely need some guidance on how to best convert this to the new HCI API (especially the DTS parts of it).

It is likely we need to have a new "zephyr,bt-hci-spi" binding. And the bt-hci dts node may need to adapt for Apollo4 Blue from:
bt-hci@0 {
	compatible = "ambiq,bt-hci-spi";
	reg = <0>;
	irq-gpios = <&gpio32_63 21 GPIO_ACTIVE_HIGH>;
	reset-gpios = <&gpio32_63 23 GPIO_ACTIVE_LOW>;
	clkreq-gpios = <&gpio32_63 20 GPIO_ACTIVE_HIGH>;
};
to:
bt_hci_spi: bt_hci_spi {
	compatible = "zephyr,bt-hci-spi";
	status = "okay";

	bt-hci@0 {
		compatible = "ambiq,bt-hci-spi";
		reg = <0>;
		irq-gpios = <&gpio32_63 21 GPIO_ACTIVE_HIGH>;
		reset-gpios = <&gpio32_63 23 GPIO_ACTIVE_LOW>;
		clkreq-gpios = <&gpio32_63 20 GPIO_ACTIVE_HIGH>;
	};
};

chosen {
	zephyr,bt-hci = &bt_hci_spi;
};

In the driver, I think we also need a hci_spi_config struct like:
struct hci_spi_config {
	struct bt_hci_info info;
	const struct device *hci_spi;
	k_thread_stack_t *rx_thread_stack;
	size_t rx_thread_stack_size;
	struct k_thread *rx_thread;
};

but not sure if we must include all global variables to one new hci_spi_data structure.
For the API adaption I think you already have a good implementation for drivers/bluetooth/hci/h4.c in https://github.com/zephyrproject-rtos/zephyr/pull/72323. The hci_ambiq.c should be similar.

Anyway, I will test the changes once you finish the upgrading on the supported Apollo3/4 EVB.

aaronyegx commented:
@RichardSWheatley Please help take another look, thanks.
